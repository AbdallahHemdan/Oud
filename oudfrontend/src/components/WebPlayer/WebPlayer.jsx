import React, { Fragment } from "react";
import { Component } from "react";
import axios from "axios";
import Player from "./Player/Player";
import Queue from "./Queue/Queue";
import Swal from "sweetalert2";
import { playback } from "./../../api/playback";
import {
  saveTrack,
  removeSavedTrack,
  setPlaying
} from "../../utils/Actions/Player";
import { base, mockUrl } from "./../../config/environment";
// import { config } from "./../../utils/auth";
const config = {
  headers: {
    authorization: `Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjVlOTA3ZGIwYTA2NDVmNDU4MTYwNzYwNiIsImlhdCI6MTU5MTYyMTQxOSwiZXhwIjoxNTk0MjEzNDE5fQ.fj3N3Pc89Pf_xlt7fGmXw1SINTecUB4-y3pihAAPjC8`,
    "Access-Control-Allow-Origin": "*"
  }
};
/**
 * Component controlling the Player and Qeueu Components. It masters all the logic behind the web player.
 * @author Ahmed Ashraf
 * @component
 * @example
 * return (
 *  <WebPlayer />
 * )
 */
class WebPlayer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      context: {},
      trackIdx: 0,
      trackId: "",
      queue: [],
      playing: false,
      loved: false
    };
    this.queueElement = React.createRef();
    this.playerElement = React.createRef();
  }
  /**
   * Axios GET request
   * @function
   * @param {string} endpoint the required endpoint to implement an action
   * @returns {object}
   */
  getRequest = endpoint => {
    return axios.get(endpoint, config);
  };
  /**
   * Axios DELETE request
   * @function
   * @param {string} endpoint the required endpoint to implement an action
   * @returns {object}
   */
  deleteRequest = endpoint => {
    return axios.delete(endpoint, config);
  };
  /**
   * Axios PUT request
   * @function
   * @param {string} endpoint the required endpoint to implement an action
   * @param {object} body the request body
   * @returns {object}
   */
  putRequest = (endpoint, body = {}) => {
    return axios.put(endpoint, body, config);
  };
  /**
   * Axios POST request
   * @function
   * @param {string} endpoint the required endpoint to implement an action
   * @param {object} body the request body
   * @returns {object}
   */
  postRequest = (endpoint, body = {}) => {
    return axios.post(endpoint, body, config);
  };
  /**
   * Axios PATCH request
   * @function
   * @param {string} endpoint the required endpoint to implement an action
   * @param {object} body the request body
   * @returns {object}
   */
  patchRequest = (endpoint, body = {}) => {
    return axios.patch(endpoint, body, config);
  };
  /**
   * A function to fetch the queue immediately afer the player has mounted.
   * @function
   * @param {number} queueIndex The index of the queue Default 0 must be greater than 0 if 0 get the current queue if 1 get the last queue
   * @param {number} trackId the id of the currently fetched track in the Player component
   * @returns {void}
   */
  fetchQueue = (queueIndex = "0", trackId = "", newQueue = false) => {
    console.log("fetching queue...");
    this.getRequest(`${base}/me/queue?queueIndex=${queueIndex}`)
      .then(response => {
        const data = response["data"];
        if (!data.hasOwnProperty("status")) {
          const size = data["total"];
          let trackIdx = 0;
          for (let i = 0; i < size; ++i)
            if (data["tracks"][i] === trackId) {
              trackIdx = i;
              break;
            }
          this.setState(
            {
              trackIdx: newQueue ? 0 : trackIdx,
              queue: data["tracks"],
              trackId: newQueue ? data["tracks"][0] : trackId
            },
            () => {
              console.log("queue:");
              console.log(this.state.queue);
              console.log(
                "idx after fetching the queue: " + this.state.trackIdx
              );
            }
          );
        }
      })
      .catch(function(error) {
        console.log(error.response);
      });
  };
  /**
   * A function to fetch any track from the server
   * @function
   * @param {number} trackId the id of the wanted track
   * @returns {object}
   */
  fetchTrack = trackId => {
    let trackIdx = this.state.trackIdx;
    for (let i = 0; i < this.state.queue.length; ++i)
      if (this.state.queue[i] === trackId) {
        trackIdx = i;
        break;
      }
    this.setState({
      trackIdx: trackIdx
    });
    return this.getRequest(`${base}/tracks/${trackId}`);
  };
  /**
   * A function to fetch the next track to the currently playing track
   * @function
   * @returns {object}
   */
  getNext = () => {
    let idx = this.state.trackIdx + 1;
    if (
      idx === this.state.queue.length &&
      this.playerElement.current.state.repeatState === 1
    ) {
      idx = 0;
    } else if (idx === this.state.queue.length) idx = idx - 1;

    this.setState({
      trackIdx: idx,
      trackId: this.state.queue[idx]
    });
    return idx;
  };
  /**
   * A function to fetch the previos track to the currently playing track
   * @function
   * @returns {object}
   */
  getPrevious = () => {
    const idx = this.state.trackIdx - 1 < 0 ? 0 : this.state.trackIdx - 1;
    this.setState({
      trackIdx: idx,
      trackId: this.state.queue[idx]
    });
    return idx;
  };
  /**
   * A generic function to play a track from any location other than the queue and the player
   * @function
   * @param {string} context OUD context
   * @param {array} uris OUD uris, list of oud tracks uris
   * @param {number} offset indext of the track in the context
   * @param {number} position from where to start in the track. In milliseconds
   * @reutrns {void}
   */
  playContext = (
    contextUri = this.state.context,
    uris = [],
    offset = 0,
    position = 0,
    isTracksList = false,
    resume = false
  ) => {
    if (base === mockUrl)
      this.putRequest(`${base}/player`, playback)
        .then(res => {
          console.log("done mocking");
        })
        .catch(err => {
          console.log("fail mock");
        });
    if (resume) {
      this.playerElement.current.handlePlayPause();
      return;
    }
    const body = isTracksList
      ? {
          uris: uris,
          offset: { position: offset }
        }
      : {
          contextUri: contextUri,
          offset: { position: offset }
        };
    this.putRequest(`${base}/me/player/play?queueIndex=0`, body)
      .then(response => {
        // this.fetchQueue();
        let player = this.playerElement.current;
        setTimeout(() => {
          player.fetchPlayback(true).then(audioUrl => {
            setTimeout(() => {
              player.playTrack(audioUrl);
            }, 100);
          });
        }, 100);
      })
      .catch(error => {
        console.log(error.response.data.message);
      });
  };
  /**
   * Handling the change in the order of the tracks in the queue.
   * @function
   * @param {array} queue the currently playing queue
   * @returns {void}
   */
  onChangeQueueOrder = (
    oldIdx = this.state.trackIdx,
    newIdx = this.state.trackIdx
  ) => {
    this.patchRequest(
      `${base}/me/queue?queueIndex=0&trackIndex=${oldIdx}&newIndex=${newIdx}`
    )
      .then(response => {
        this.fetchQueue(0, this.state.trackId);
      })
      .catch(error => {
        console.log(error.response);
      });
  };
  /**
   * Change the playing state as there are many stuff depend on that in the UI
   * @function
   * @param {boot} playing playing state of the track
   * @returns {void}
   */
  changePlayingState = (
    playing,
    idx = this.state.trackIdx,
    id = this.state.trackId
  ) => {
    this.setState({
      playing: playing,
      trackIdx: idx,
      trackId: id
    });
  };
  togglePlayingState = () => {
    this.setState({
      playing: !this.state.playing
    });
    return this.state.playing;
  };
  changeLovedState = state => {
    this.setState({
      loved: state
    });
  };
  /**
   * Hadling the removing of a track from the queue
   * @function
   * @param {number} idx the index of the track in the queue
   * @param {string} id the unique id of the track
   * @returns {void}
   */
  removeTrack = (idx, id) => {
    //
    this.deleteRequest(`${base}/me/queue?trackId=${id}`)
      .then(response => {
        this.fetchQueue(0, this.state.trackId);
        Swal.fire({
          title: "Done!",
          text: "Track Deleted Successfully from your Queue!",
          icon: "success",
          showConfirmButton: false,
          timer: 1000
        });
      })
      .catch(error => {
        console.log(error.response);
      });
  };
  likeSong = (trackId, queue = false) => {
    saveTrack(trackId).then(done => {
      if (done) {
        this.setState({
          loved: queue ? this.state.loved : true
        });
        Swal.fire({
          title: "Done!",
          text: "Added to your Liked Songs!",
          icon: "success",
          showConfirmButton: false,
          timer: 1000
        });
      }
    });
  };
  unlikeSong = (trackId, queue = false) => {
    removeSavedTrack(trackId)
      .then(done => {
        if (done || base === mockUrl)
          this.setState({
            loved: queue ? this.state.loved : false
          });
        Swal.fire({
          title: "Done!",
          text: "Removed from your Liked Songs!",
          icon: "success",
          showConfirmButton: false,
          timer: 1000
        });
      })
      .catch(error => {
        console.log(error.response.data.message);
      });
  };
  render() {
    return (
      <Fragment>
        <Queue
          ref={this.queueElement}
          tracks={this.state.queue}
          trackId={this.state.trackId}
          trackIdx={this.state.trackIdx}
          deviceId={this.state.deviceId}
          playing={this.state.playing}
          changePlayingState={this.changePlayingState}
          onChangeQueueOrder={this.onChangeQueueOrder}
          player={this.playerElement}
          removeTrack={this.removeTrack}
          likeSong={this.likeSong}
          unlikeSong={this.unlikeSong}
          data-testid="queue-container"
        />
        <Player
          ref={this.playerElement}
          trackId={this.state.trackId}
          trackIdx={this.state.trackIdx}
          playing={this.state.playing}
          queueElement={this.queueElement}
          getRequest={this.getRequest}
          putRequest={this.putRequest}
          postRequest={this.postRequest}
          fetchQueue={this.fetchQueue}
          getNext={this.getNext}
          getPrevious={this.getPrevious}
          changePlayingState={this.changePlayingState}
          fetchTrack={this.fetchTrack}
          likeSong={this.likeSong}
          unlikeSong={this.unlikeSong}
          loved={this.state.loved}
          changeLovedState={this.changeLovedState}
          data-testid="player-container"
        />
      </Fragment>
    );
  }
}

export default WebPlayer;
