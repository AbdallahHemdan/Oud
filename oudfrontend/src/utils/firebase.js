import firebase from "firebase";

const firebasConfig = {
  apiKey: "AIzaSyBZVUambAFEB-ywzAZTAPP-tSkbYXSEWc4",
  authDomain: "oud-notifications.firebaseapp.com",
  databaseURL: "https://oud-notifications.firebaseio.com",
  projectId: "oud-notifications",
  storageBucket: "oud-notifications.appspot.com",
  messagingSenderId: "681943816425",
  appId: "1:681943816425:web:9149a8834b2ff0f4623878"
};
// Initialize Firebase
firebase.initializeApp(firebasConfig);

export default firebase;
