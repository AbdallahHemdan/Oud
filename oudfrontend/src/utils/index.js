import axios from 'axios'
import { base } from '../config/environment'
const findByTestAttr = (component, val) => {
    return component.find(`[data-testid="${val}"]`);
}

function addToLikedSongs(id) {
    axios.put(`${base}/me/tracks/${id}`)
        .then(function (response) {
            console.log(response);

        })
        .catch(function (error) {
            console.log(error);
        });
}
function removeLikedSong(id) {
    axios.delete(`${base}/me/tracks/${id}`)
        .then(function (response) {
            console.log(response);
        })
        .catch(function (error) {
            console.log(error);
        });
}
export { findByTestAttr, addToLikedSongs, removeLikedSong }
