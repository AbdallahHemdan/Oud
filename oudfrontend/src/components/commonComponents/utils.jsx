import axios from "axios";
import { base } from "../../config/environment"
import {config} from "../../utils/auth"
/**
 * @file this file contains three functions used in Playlist, Album, LikedSongs 
 **/
export function addSong(playlistId, track){
  axios
  .post(`${base}/playlists/${playlistId}/${track}`, config)
  .then((response) => {
  })
  .catch((error) => {
  });
}