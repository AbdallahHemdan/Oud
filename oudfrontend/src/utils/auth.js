import jwtDecode from "jwt-decode";
import { getRequest } from "./requester";
import { base } from "./../config/environment";

function getToken() {
  return localStorage.getItem("accessToken");
}

const config = {
  headers: { authorization: `Bearer ${getToken()}` }
};

function Auth() {
  try {
    if (!getToken()) return null;
    let decoded = jwtDecode(getToken());
    console.log(decoded.id);
    return decoded.id;
  } catch {
    return null;
  }
}
function isLoggedIn() {
  return getToken() ? true : false;
}

function isArtist() {
  return getRequest(`${base}/me`)
    .then(response => {
      return response.data.role === "artist";
    })
    .catch(error => {
      console.log(error);
    });
}
const _config = config;
export { _config as config };
export const token = getToken();
const _Auth = Auth;
export { _Auth as Auth };
const _isLoggedIn = isLoggedIn;
export { _isLoggedIn as isLoggedIn };
const _isArtist = isArtist;
export { _isArtist as isArtist };
