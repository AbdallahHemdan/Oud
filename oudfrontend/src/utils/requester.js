import axios from "axios";
import { isLoggedIn } from "./../utils/auth";
const config = {
  headers: {
    authorization: `Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjVlOTA3ZGIwYTA2NDVmNDU4MTYwNzYwNiIsImlhdCI6MTU5MTYyMTQxOSwiZXhwIjoxNTk0MjEzNDE5fQ.fj3N3Pc89Pf_xlt7fGmXw1SINTecUB4-y3pihAAPjC8`,
    "Access-Control-Allow-Origin": "*"
  }
};
const getRequest = endpoint => {
  return axios.get(endpoint, config);
};
/**
 * Axios DELETE request
 * @function
 * @param {string} endpoint the required endpoint to implement an action
 * @returns {object}
 */
const deleteRequest = endpoint => {
  return axios.delete(endpoint, config);
};
/**
 * Axios PUT request
 * @function
 * @param {string} endpoint the required endpoint to implement an action
 * @param {object} body the request body
 * @returns {object}
 */
const putRequest = (endpoint, body = {}) => {
  return axios.put(endpoint, body, config);
};
/**
 * Axios POST request
 * @function
 * @param {string} endpoint the required endpoint to implement an action
 * @param {object} body the request body
 * @returns {object}
 */
const postRequest = (endpoint, body = {}) => {
  return axios.post(endpoint, body, config);
};
/**
 * Axios PATCH request
 * @function
 * @param {string} endpoint the required endpoint to implement an action
 * @param {object} body the request body
 * @returns {object}
 */
const patchRequest = (endpoint, body = {}) => {
  return axios.patch(endpoint, body, config);
};

// const isArtist = () => {
//   if (isLoggedIn()) {
//   }
// };
export { getRequest, deleteRequest, putRequest, postRequest, patchRequest };
