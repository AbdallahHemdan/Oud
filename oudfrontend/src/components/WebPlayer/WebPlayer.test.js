import React from "react";
import { shallow, mount } from "enzyme";
import WebPlayer from "./WebPlayer";
import { findByTestAttr } from "./../../utils/index";

const setup = () => {
  const component = shallow(<WebPlayer />);
  return component;
};

describe("Web Player Component", () => {
  it("snapshot testing", () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it("check redering queue correctly", () => {
    const wrapper = findByTestAttr(setup(), "queue-container");
    expect(wrapper.length).toBe(1);
  });

  it("check redering player correctly", () => {
    const wrapper = findByTestAttr(setup(), "queue-container");
    expect(wrapper.length).toBe(1);
  });

  it("check change love state", () => {
    const wrapper = mount(<WebPlayer />);
    wrapper.instance().changeLovedState(true);
    expect(wrapper.state().loved).toBe(true);
  });

  it("check toggle playing state", () => {
    const wrapper = mount(<WebPlayer />);
    const playing = wrapper.state().playing;
    const newPlaying = wrapper.instance().togglePlayingState();
    expect(newPlaying).toBe(!playing);
  });

  it("check change playing, idx, and id state", () => {
    const wrapper = mount(<WebPlayer />);
    wrapper.instance().changePlayingState(true, 5, "sjdfhye54dcsc1");
    expect(wrapper.state().playing).toBe(true);
    expect(wrapper.state().trackIdx).toBe(5);
    expect(wrapper.state().trackId).toBe("sjdfhye54dcsc1");
  });

  it("check getPrvious changes index correctly with zero index", () => {
    const wrapper = mount(<WebPlayer />);
    wrapper.instance().setState({
      trackIdx: 0,
    });
    const idx = wrapper.instance().getPrevious();
    expect(idx).toBe(0);
  });

  it("check getPrvious changes index correctly with non-zero index", () => {
    const wrapper = mount(<WebPlayer />);
    wrapper.instance().setState({
      trackIdx: 5,
    });
    const idx = wrapper.instance().getPrevious();
    expect(idx).toBe(4);
  });

  it("check getNext changes index correctly with end of queue, and repeat context", () => {
    const wrapper = mount(<WebPlayer />);
    wrapper.instance().setState({
      queue: [1, 2, 3],
      trackIdx: 2,
    });
    wrapper.instance().playerElement.current.setState({
      repeatState: 1,
    });
    const idx = wrapper.instance().getNext();
    expect(idx).toBe(0);
  });

  it("check getNext changes index correctly with end of queue, and repeat is off", () => {
    const wrapper = mount(<WebPlayer />);
    wrapper.instance().setState({
      queue: [1, 2, 3],
      trackIdx: 2,
    });
    wrapper.instance().playerElement.current.setState({
      repeatState: 0,
    });
    const idx = wrapper.instance().getNext();
    expect(idx).toBe(2);
  });

  it("check getNext changes index correctly with end of queue, and repeat track", () => {
    const wrapper = mount(<WebPlayer />);
    wrapper.instance().setState({
      queue: [1, 2, 3],
      trackIdx: 2,
    });
    wrapper.instance().playerElement.current.setState({
      repeatState: 2,
    });
    const idx = wrapper.instance().getNext();
    expect(idx).toBe(2);
  });

  it("check getNext changes index correctly with end of queue, normal increase", () => {
    const wrapper = mount(<WebPlayer />);
    wrapper.instance().setState({
      queue: [1, 2, 3],
      trackIdx: 1,
    });
    wrapper.instance().playerElement.current.setState({
      repeatState: 1,
    });
    const idx = wrapper.instance().getNext();
    expect(idx).toBe(2);
  });
});
