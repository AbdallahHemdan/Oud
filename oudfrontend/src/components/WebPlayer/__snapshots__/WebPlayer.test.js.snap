// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Web Player Component snapshot testing 1`] = `ShallowWrapper {}`;
